python/ql/lib/semmle/python/dataflow/new/internal/ImportResolution.qll

```diff
@@ -1,14 +1,72 @@
+/**
+ * INTERNAL. DO NOT USE.
+ *
+ * Provides predicates for resolving imports.
+ */
+
 private import python
 private import semmle.python.dataflow.new.DataFlow
 private import semmle.python.dataflow.new.internal.ImportStar
 private import semmle.python.dataflow.new.TypeTracker
 
+/**
+ * Python modules and the way imports are resolved are... complicated. Here's a crash course in how
+ * it works, as well as some caveats to bear in mind when looking at the implementation in this
+ * module.
+ *
+ * First, let's consider the humble `import` statement:
+ * ```python
+ * import foo
+ * import bar.baz
+ * import ham.eggs as spam
+ * ```
+ *
+ * In the AST, all imports are aliased, as in the last import above. That is, `import foo` becomes
+ * `import foo as foo`, and `import bar.baz` becomes `import bar as bar`. Note that `import` is
+ * exclusively used to import modules -- if `eggs` is an attribute of the `ham` module (and not a
+ * submodule of the `ham` package), then the third line above is an error.
+ *
+ * Next, we have the `from` statement. This one is a bit more complicated, but still has the same
+ * aliasing desugaring as above applied to it. Thus, `from foo import bar` becomes
+ * `from foo import bar as bar`.
+ *
+ * In general, `from foo import bar` can mean two different things:
+ *
+ * 1. If `foo` is a module, and `bar` is an attribute of `foo`, then `from foo import bar` imports
+ *    the attribute `bar` into the current module (binding it to the name `bar`).
+ * 2. If `foo` is a package, and `bar` is a submodule of `foo`, then `from foo import bar` first imports
+ *    `foo.bar`, and then attempts to locate the `bar` attribute again. In most cases, that attribute
+ *    will then point to the `bar` submodule.
```

Comment:
I'm confused about the "attempt" part, could we use the following instead?

```suggestion
 * 2. If `foo` is a package, and `bar` is a submodule of `foo`, then `from foo import bar` first imports
 *    `foo.bar`, and then reads the `bar` attribute on `foo`. In most cases, that attribute
 *    will then point to the `bar` submodule.
```

_I guess you could actually use `del bar` in `foo/__init__.py` to remove the attribute :thinking:_

AHA, I remembered about some import tests I added in https://github.com/github/codeql/commit/0265f263018f58305966d37e690a991d04960939 (but required me looking over all files I had ever changed in this repo, since I couldn't remember the new location, and couldn't find it from just searching :shrug:)

test2 is the interesting one here (I did not annotate the original testfile, but checked out the old commit to run it, and do so here).

```py
from mypkg import foo, bar
print(foo) # 42
print(bar) #<module 'mypkg.bar' ...
```

these results means that it doesn't look like the semantics are to do an `import mypkg.foo` first, and then look up the attribute, but to see if an attribute is already available on the package, if it is, use that, if it isn't, see if there is a submodule with the right name, import that as an attribute of the parent package, and then do the attribute read. After writing that, I went and looked up [the Python documentation on `from`](https://docs.python.org/3.10/reference/simple_stmts.html#index-34), and that states the same.

Could we maybe expand the tests you have created to also have such a setup? and also update our QLDocs with the right semantics?

(personally, my guess is it will not matter that we have this slightly wrong behavior, but I think it's good to have a test to show we know about it)


Here is my take on what the docs could say:

```suggestion
 * 2. If `foo` is a package, and `bar` is already defined in `foo/__init__.py`,
 *    that value will be imported. If it is not defined, and `bar` is a submodule of `foo`, then `bar` is imported to `foo`, and the `bar` submodule imported. 
 *    Note: We don't currently model if the attribute is already defined in `__init__.py`
 *    and always assume that the submodule will be used.
```

### Where are the old import tests now?

I note that since these were ported to ApiGraph in https://github.com/github/codeql/commit/e4c3544a3f2706625f4c961a38c5d9c8700486e1
the original copy was removed in https://github.com/github/codeql/commit/79c3ccd56e0ac80cbc14bb887dc3b78aafe58d01, but sadly the current copy in the repo has been made non-executable by https://github.com/github/codeql/commit/3801a158a84a89583a1f574b1f3fe4fffaaaec5a

I honestly wasn't aware this had happened, and it would be nice to get these import tests to work again, but I'm a little in over my head in regards to other tasks, so I'm not going to work on this now.